File: src/LTAdmin/Repositories/AdminRepository.cs
using System.Data;
using LTAdmin.Data;
using LTAdmin.Models.Entities;

namespace LTAdmin.Repositories;

/// <summary>Dépôt du bloc Administration (utilisateurs, paramètres, établissement, années, journal).</summary>
public sealed class AdminRepository : RepositoryBase
{
    public AdminRepository(AccessDatabase database) : base(database)
    {
    }

    // ----- UTILISATEUR -----

    public Utilisateur? GetUtilisateur(string codeUtr)
        => QuerySingle(
            $"SELECT * FROM {Q(Tables.Utilisateur)} WHERE {Q("CODE_UTR")} = ?",
            MapUtilisateur, P(codeUtr));

    public List<Utilisateur> ListUtilisateurs()
        => QueryList(
            $"SELECT * FROM {Q(Tables.Utilisateur)} ORDER BY {Q("CODE_UTR")}",
            MapUtilisateur);

    public void InsertUtilisateur(Utilisateur user)
    {
        Execute(
            $"INSERT INTO {Q(Tables.Utilisateur)} ({Q("CODE_UTR")}, {Q("NOM_UTR")}, {Q("MOT_PASSE")}, {Q("PROFIL")}, {Q("ACTIF")}) VALUES (?, ?, ?, ?, ?)",
            P(user.CodeUtr), P(user.NomUtr), P(user.MotPasse), P(user.Profil), P(user.Actif ?? true));
    }

    public int UpdateUtilisateur(Utilisateur user)
        => Execute(
            $"UPDATE {Q(Tables.Utilisateur)} SET {Q("NOM_UTR")} = ?, {Q("PROFIL")} = ?, {Q("ACTIF")} = ? WHERE {Q("CODE_UTR")} = ?",
            P(user.NomUtr), P(user.Profil), P(user.Actif ?? true), P(user.CodeUtr));

    public int UpdateMotPasse(string codeUtr, string nouveauMotPasse)
        => Execute(
            $"UPDATE {Q(Tables.Utilisateur)} SET {Q("MOT_PASSE")} = ? WHERE {Q("CODE_UTR")} = ?",
            P(nouveauMotPasse), P(codeUtr));

    public int SetUtilisateurActif(string codeUtr, bool actif)
        => Execute(
            $"UPDATE {Q(Tables.Utilisateur)} SET {Q("ACTIF")} = ? WHERE {Q("CODE_UTR")} = ?",
            P(actif), P(codeUtr));

    public int DeleteUtilisateur(string codeUtr)
        => Execute(
            $"DELETE FROM {Q(Tables.Utilisateur)} WHERE {Q("CODE_UTR")} = ?",
            P(codeUtr));

    // ----- PARAMETRE -----

    public List<Parametre> ListParametres()
        => QueryList(
            $"SELECT * FROM {Q(Tables.Parametre)} ORDER BY {Q("CLE")}",
            MapParametre);

    public string? GetValeur(string cle)
        => ScalarString(
            $"SELECT {Q("VALEUR")} FROM {Q(Tables.Parametre)} WHERE {Q("CLE")} = ?",
            P(cle));

    public int UpdateValeur(string cle, string valeur)
        => Execute(
            $"UPDATE {Q(Tables.Parametre)} SET {Q("VALEUR")} = ? WHERE {Q("CLE")} = ?",
            P(valeur), P(cle));

    // ----- ETABLISSEMENT -----

    public Etablissement? GetEtablissement()
        => QuerySingle(
            $"SELECT TOP 1 * FROM {Q(Tables.Etablissement)} ORDER BY {Q("CODE_ETAB")}",
            MapEtablissement);

    public int UpdateEtablissement(Etablissement etablissement)
        => Execute(
            $"UPDATE {Q(Tables.Etablissement)} SET {Q("NOM_ETAB")} = ?, {Q("SIGLE")} = ?, {Q("ADRESSE")} = ?, " +
            $"{Q("TEL")} = ?, {Q("EMAIL")} = ?, {Q("SITE_WEB")} = ?, {Q("DIRECTEUR")} = ?, {Q("LOGO")} = ? " +
            $"WHERE {Q("CODE_ETAB")} = ?",
            P(etablissement.NomEtab), P(etablissement.Sigle), P(etablissement.Adresse),
            P(etablissement.Tel), P(etablissement.Email), P(etablissement.SiteWeb),
            P(etablissement.Directeur), P(etablissement.Logo), P(etablissement.CodeEtab));

    // ----- ANNEE_SCOLAIRE -----

    public List<AnneeScolaire> ListAnnees()
        => QueryList(
            $"SELECT * FROM {Q(Tables.AnneeScolaire)} ORDER BY {Q("LIBELLE")} DESC",
            MapAnnee);

    public AnneeScolaire? GetAnnee(int idAnnee)
        => QuerySingle(
            $"SELECT * FROM {Q(Tables.AnneeScolaire)} WHERE {Q("ID_ANNEE")} = ?",
            MapAnnee, P(idAnnee));

    public AnneeScolaire? GetAnneeActive()
        => QuerySingle(
            $"SELECT TOP 1 * FROM {Q(Tables.AnneeScolaire)} WHERE {Q("ACTIVE")} = ? ORDER BY {Q("ID_ANNEE")} DESC",
            MapAnnee, P(true));

    public int InsertAnnee(AnneeScolaire annee)
        => InsertAndGetId(
            $"INSERT INTO {Q(Tables.AnneeScolaire)} ({Q("LIBELLE")}, {Q("DATE_DEBUT")}, {Q("DATE_FIN")}, {Q("ACTIVE")}) VALUES (?, ?, ?, ?)",
            P(annee.Libelle), P(annee.DateDebut), P(annee.DateFin), P(annee.Active ?? false));

    public int UpdateAnnee(AnneeScolaire annee)
        => Execute(
            $"UPDATE {Q(Tables.AnneeScolaire)} SET {Q("LIBELLE")} = ?, {Q("DATE_DEBUT")} = ?, {Q("DATE_FIN")} = ?, {Q("ACTIVE")} = ? WHERE {Q("ID_ANNEE")} = ?",
            P(annee.Libelle), P(annee.DateDebut), P(annee.DateFin), P(annee.Active ?? false), P(annee.IdAnnee));

    /// <summary>Active une année et désactive les autres, en transaction.</summary>
    public void SetAnneeActive(int idAnnee)
    {
        using var tx = Db.BeginTransaction();
        Execute($"UPDATE {Q(Tables.AnneeScolaire)} SET {Q("ACTIVE")} = ?", P(false));
        Execute($"UPDATE {Q(Tables.AnneeScolaire)} SET {Q("ACTIVE")} = ? WHERE {Q("ID_ANNEE")} = ?", P(true), P(idAnnee));
        tx.Complete();
    }

    // ----- JOURNAL -----

    public int InsertJournal(JournalEntry entry)
        => InsertAndGetId(
            $"INSERT INTO {Q(Tables.Journal)} ({Q("DATE_LOG")}, {Q("CODE_UTR")}, {Q("ACTION_LOG")}, {Q("TABLE_CIBLE")}, {Q("ID_CIBLE")}, {Q("DETAIL")}) VALUES (?, ?, ?, ?, ?, ?)",
            P(entry.DateLog ?? DateTime.Now), P(entry.CodeUtr), P(entry.ActionLog),
            P(entry.TableCible), P(entry.IdCible), P(entry.Detail));

    public List<JournalEntry> ListJournal(DateTime? depuis, string? codeUtr, string? action, int maxRows)
    {
        var sql = $"SELECT TOP {Math.Clamp(maxRows, 1, 5000)} * FROM {Q(Tables.Journal)}";
        var conditions = new List<string>();
        var parameters = new List<System.Data.OleDb.OleDbParameter>();
        if (depuis.HasValue) { conditions.Add($"{Q("DATE_LOG")} >= ?"); parameters.Add(P(depuis.Value)); }
        if (!string.IsNullOrWhiteSpace(codeUtr)) { conditions.Add($"{Q("CODE_UTR")} = ?"); parameters.Add(P(codeUtr.Trim())); }
        if (!string.IsNullOrWhiteSpace(action)) { conditions.Add($"{Q("ACTION_LOG")} = ?"); parameters.Add(P(action.Trim())); }
        if (conditions.Count > 0) sql += " WHERE " + string.Join(" AND ", conditions);
        sql += $" ORDER BY {Q("ID_LOG")} DESC";
        return QueryList(sql, MapJournal, parameters.ToArray());
    }

    public int CountJournal() => ScalarInt($"SELECT COUNT(*) FROM {Q(Tables.Journal)}");

    // ----- Mappers -----

    public static Utilisateur MapUtilisateur(DataRow row) => new()
    {
        CodeUtr = DataRowMapper.GetString(row, "CODE_UTR"),
        NomUtr = DataRowMapper.GetString(row, "NOM_UTR"),
        MotPasse = DataRowMapper.GetString(row, "MOT_PASSE"),
        Profil = DataRowMapper.GetString(row, "PROFIL"),
        Actif = DataRowMapper.GetBoolean(row, "ACTIF")
    };

    public static Parametre MapParametre(DataRow row) => new()
    {
        Cle = DataRowMapper.GetString(row, "CLE"),
        Valeur = DataRowMapper.GetString(row, "VALEUR"),
        Description = DataRowMapper.GetString(row, "DESCRIPTION_P")
    };

    public static Etablissement MapEtablissement(DataRow row) => new()
    {
        CodeEtab = DataRowMapper.GetString(row, "CODE_ETAB"),
        NomEtab = DataRowMapper.GetString(row, "NOM_ETAB"),
        Sigle = DataRowMapper.GetString(row, "SIGLE"),
        Adresse = DataRowMapper.GetString(row, "ADRESSE"),
        Tel = DataRowMapper.GetString(row, "TEL"),
        Email = DataRowMapper.GetString(row, "EMAIL"),
        SiteWeb = DataRowMapper.GetString(row, "SITE_WEB"),
        Directeur = DataRowMapper.GetString(row, "DIRECTEUR"),
        Logo = DataRowMapper.GetString(row, "LOGO")
    };

    public static AnneeScolaire MapAnnee(DataRow row) => new()
    {
        IdAnnee = DataRowMapper.GetInt32(row, "ID_ANNEE"),
        Libelle = DataRowMapper.GetString(row, "LIBELLE"),
        DateDebut = DataRowMapper.GetDateTime(row, "DATE_DEBUT"),
        DateFin = DataRowMapper.GetDateTime(row, "DATE_FIN"),
        Active = DataRowMapper.GetBoolean(row, "ACTIVE")
    };

    public static JournalEntry MapJournal(DataRow row) => new()
    {
        IdLog = DataRowMapper.GetInt32(row, "ID_LOG"),
        DateLog = DataRowMapper.GetDateTime(row, "DATE_LOG"),
        CodeUtr = DataRowMapper.GetString(row, "CODE_UTR"),
        ActionLog = DataRowMapper.GetString(row, "ACTION_LOG"),
        TableCible = DataRowMapper.GetString(row, "TABLE_CIBLE"),
        IdCible = DataRowMapper.GetInt32(row, "ID_CIBLE"),
        Detail = DataRowMapper.GetString(row, "DETAIL")
    };
}

File: ltadmin/repositories/admin_repository.py
"""Dépôt du bloc Administration (utilisateurs, paramètres, établissement, années, journal)."""

from __future__ import annotations

import datetime as _dt
from typing import List, Optional

from ltadmin.data import row_mapper as rm
from ltadmin.data.schema import Tables
from ltadmin.models.entities import (
    AnneeScolaire,
    Etablissement,
    JournalEntry,
    Parametre,
    Utilisateur,
)
from ltadmin.repositories.base import RepositoryBase, p, q


class AdminRepository(RepositoryBase):

    # ----- UTILISATEUR -----

    def get_utilisateur(self, code_utr: str) -> Optional[Utilisateur]:
        # Access compare les chaînes sans tenir compte de la casse par défaut ;
        # UCASE rend ce comportement explicite et identique sous SQLite.
        return self.query_single(
            f"SELECT * FROM {q(Tables.UTILISATEUR)} "
            f"WHERE UCASE({q('CODE_UTR')}) = UCASE(?)",
            UtilisateurMapper.map, [p(code_utr)])

    def list_utilisateurs(self) -> List[Utilisateur]:
        return self.query_list(
            f"SELECT * FROM {q(Tables.UTILISATEUR)} ORDER BY {q('CODE_UTR')}",
            UtilisateurMapper.map)

    def insert_utilisateur(self, user: Utilisateur) -> None:
        self.execute(
            f"INSERT INTO {q(Tables.UTILISATEUR)} ({q('CODE_UTR')}, {q('NOM_UTR')}, "
            f"{q('MOT_PASSE')}, {q('PROFIL')}, {q('ACTIF')}) VALUES (?, ?, ?, ?, ?)",
            [p(user.code_utr), p(user.nom_utr), p(user.mot_passe), p(user.profil),
             p(True if user.actif is None else user.actif)])

    def update_utilisateur(self, user: Utilisateur) -> int:
        return self.execute(
            f"UPDATE {q(Tables.UTILISATEUR)} SET {q('NOM_UTR')} = ?, {q('PROFIL')} = ?, "
            f"{q('ACTIF')} = ? WHERE {q('CODE_UTR')} = ?",
            [p(user.nom_utr), p(user.profil), p(True if user.actif is None else user.actif),
             p(user.code_utr)])

    def update_mot_passe(self, code_utr: str, nouveau_mot_passe: str) -> int:
        return self.execute(
            f"UPDATE {q(Tables.UTILISATEUR)} SET {q('MOT_PASSE')} = ? WHERE {q('CODE_UTR')} = ?",
            [p(nouveau_mot_passe), p(code_utr)])

    def set_utilisateur_actif(self, code_utr: str, actif: bool) -> int:
        return self.execute(
            f"UPDATE {q(Tables.UTILISATEUR)} SET {q('ACTIF')} = ? WHERE {q('CODE_UTR')} = ?",
            [p(actif), p(code_utr)])

    def delete_utilisateur(self, code_utr: str) -> int:
        return self.execute(
            f"DELETE FROM {q(Tables.UTILISATEUR)} WHERE {q('CODE_UTR')} = ?", [p(code_utr)])

    # ----- PARAMETRE -----

    def list_parametres(self) -> List[Parametre]:
        return self.query_list(
            f"SELECT * FROM {q(Tables.PARAMETRE)} ORDER BY {q('CLE')}",
            AdminRepository.map_parametre)

    def get_valeur(self, cle: str) -> Optional[str]:
        return self.scalar_string(
            f"SELECT {q('VALEUR')} FROM {q(Tables.PARAMETRE)} WHERE {q('CLE')} = ?", [p(cle)])

    def update_valeur(self, cle: str, valeur: str) -> int:
        return self.execute(
            f"UPDATE {q(Tables.PARAMETRE)} SET {q('VALEUR')} = ? WHERE {q('CLE')} = ?",
            [p(valeur), p(cle)])

    # ----- ETABLISSEMENT -----

    def get_etablissement(self) -> Optional[Etablissement]:
        return self.query_single(
            f"SELECT TOP 1 * FROM {q(Tables.ETABLISSEMENT)} ORDER BY {q('CODE_ETAB')}",
            AdminRepository.map_etablissement)

    def update_etablissement(self, etablissement: Etablissement) -> int:
        return self.execute(
            f"UPDATE {q(Tables.ETABLISSEMENT)} SET {q('NOM_ETAB')} = ?, {q('SIGLE')} = ?, "
            f"{q('ADRESSE')} = ?, {q('TEL')} = ?, {q('EMAIL')} = ?, {q('SITE_WEB')} = ?, "
            f"{q('DIRECTEUR')} = ?, {q('LOGO')} = ? WHERE {q('CODE_ETAB')} = ?",
            [p(etablissement.nom_etab), p(etablissement.sigle), p(etablissement.adresse),
             p(etablissement.tel), p(etablissement.email), p(etablissement.site_web),
             p(etablissement.directeur), p(etablissement.logo), p(etablissement.code_etab)])

    # ----- ANNEE_SCOLAIRE -----

    def list_annees(self) -> List[AnneeScolaire]:
        return self.query_list(
            f"SELECT * FROM {q(Tables.ANNEE_SCOLAIRE)} ORDER BY {q('LIBELLE')} DESC",
            AdminRepository.map_annee)

    def get_annee(self, id_annee: int) -> Optional[AnneeScolaire]:
        return self.query_single(
            f"SELECT * FROM {q(Tables.ANNEE_SCOLAIRE)} WHERE {q('ID_ANNEE')} = ?",
            AdminRepository.map_annee, [p(id_annee)])

    def get_annee_active(self) -> Optional[AnneeScolaire]:
        return self.query_single(
            f"SELECT TOP 1 * FROM {q(Tables.ANNEE_SCOLAIRE)} WHERE {q('ACTIVE')} = ? "
            f"ORDER BY {q('ID_ANNEE')} DESC",
            AdminRepository.map_annee, [p(True)])

    def insert_annee(self, annee: AnneeScolaire) -> int:
        return self.insert_and_get_id(
            f"INSERT INTO {q(Tables.ANNEE_SCOLAIRE)} ({q('LIBELLE')}, {q('DATE_DEBUT')}, "
            f"{q('DATE_FIN')}, {q('ACTIVE')}) VALUES (?, ?, ?, ?)",
            [p(annee.libelle), p(annee.date_debut), p(annee.date_fin),
             p(False if annee.active is None else annee.active)])

    def update_annee(self, annee: AnneeScolaire) -> int:
        return self.execute(
            f"UPDATE {q(Tables.ANNEE_SCOLAIRE)} SET {q('LIBELLE')} = ?, {q('DATE_DEBUT')} = ?, "
            f"{q('DATE_FIN')} = ?, {q('ACTIVE')} = ? WHERE {q('ID_ANNEE')} = ?",
            [p(annee.libelle), p(annee.date_debut), p(annee.date_fin),
             p(False if annee.active is None else annee.active), p(annee.id_annee)])

    def set_annee_active(self, id_annee: int) -> None:
        """Active une année et désactive les autres, en transaction."""
        with self.db.begin_transaction():
            self.execute(f"UPDATE {q(Tables.ANNEE_SCOLAIRE)} SET {q('ACTIVE')} = ?", [p(False)])
            self.execute(
                f"UPDATE {q(Tables.ANNEE_SCOLAIRE)} SET {q('ACTIVE')} = ? WHERE {q('ID_ANNEE')} = ?",
                [p(True), p(id_annee)])

    # ----- JOURNAL -----

    def insert_journal(self, entry: JournalEntry) -> int:
        return self.insert_and_get_id(
            f"INSERT INTO {q(Tables.JOURNAL)} ({q('DATE_LOG')}, {q('CODE_UTR')}, "
            f"{q('ACTION_LOG')}, {q('TABLE_CIBLE')}, {q('ID_CIBLE')}, {q('DETAIL')}) "
            f"VALUES (?, ?, ?, ?, ?, ?)",
            [p(entry.date_log or _dt.datetime.now()), p(entry.code_utr), p(entry.action_log),
             p(entry.table_cible), p(entry.id_cible), p(entry.detail)])

    def list_journal(self, depuis=None, code_utr: Optional[str] = None,
                     action: Optional[str] = None, max_rows: int = 500) -> List[JournalEntry]:
        max_rows = max(1, min(max_rows, 5000))
        sql = f"SELECT TOP {max_rows} * FROM {q(Tables.JOURNAL)}"
        conditions = []
        parameters = []
        if depuis is not None:
            conditions.append(f"{q('DATE_LOG')} >= ?")
            parameters.append(p(depuis))
        if code_utr and code_utr.strip():
            conditions.append(f"{q('CODE_UTR')} = ?")
            parameters.append(p(code_utr.strip()))
        if action and action.strip():
            conditions.append(f"{q('ACTION_LOG')} = ?")
            parameters.append(p(action.strip()))
        if conditions:
            sql += " WHERE " + " AND ".join(conditions)
        sql += f" ORDER BY {q('ID_LOG')} DESC"
        return self.query_list(sql, AdminRepository.map_journal, parameters)

    def count_journal(self) -> int:
        return self.scalar_int(f"SELECT COUNT(*) FROM {q(Tables.JOURNAL)}")

    # ----- Mappers -----

    @staticmethod
    def map_utilisateur(row: dict) -> Utilisateur:
        return Utilisateur(
            code_utr=rm.get_string(row, "CODE_UTR"),
            nom_utr=rm.get_string(row, "NOM_UTR"),
            mot_passe=rm.get_string(row, "MOT_PASSE"),
            profil=rm.get_string(row, "PROFIL"),
            actif=rm.get_bool(row, "ACTIF"),
        )

    @staticmethod
    def map_parametre(row: dict) -> Parametre:
        return Parametre(
            cle=rm.get_string(row, "CLE"),
            valeur=rm.get_string(row, "VALEUR"),
            description=rm.get_string(row, "DESCRIPTION_P"),
        )

    @staticmethod
    def map_etablissement(row: dict) -> Etablissement:
        return Etablissement(
            code_etab=rm.get_string(row, "CODE_ETAB"),
            nom_etab=rm.get_string(row, "NOM_ETAB"),
            sigle=rm.get_string(row, "SIGLE"),
            adresse=rm.get_string(row, "ADRESSE"),
            tel=rm.get_string(row, "TEL"),
            email=rm.get_string(row, "EMAIL"),
            site_web=rm.get_string(row, "SITE_WEB"),
            directeur=rm.get_string(row, "DIRECTEUR"),
            logo=rm.get_string(row, "LOGO"),
        )

    @staticmethod
    def map_annee(row: dict) -> AnneeScolaire:
        return AnneeScolaire(
            id_annee=rm.get_int(row, "ID_ANNEE"),
            libelle=rm.get_string(row, "LIBELLE"),
            date_debut=rm.get_datetime(row, "DATE_DEBUT"),
            date_fin=rm.get_datetime(row, "DATE_FIN"),
            active=rm.get_bool(row, "ACTIVE"),
        )

    @staticmethod
    def map_journal(row: dict) -> JournalEntry:
        return JournalEntry(
            id_log=rm.get_int(row, "ID_LOG"),
            date_log=rm.get_datetime(row, "DATE_LOG"),
            code_utr=rm.get_string(row, "CODE_UTR"),
            action_log=rm.get_string(row, "ACTION_LOG"),
            table_cible=rm.get_string(row, "TABLE_CIBLE"),
            id_cible=rm.get_int(row, "ID_CIBLE"),
            detail=rm.get_string(row, "DETAIL"),
        )


# Alias pratique (le C# exposait MapUtilisateur en statique public).
class UtilisateurMapper:
    map = AdminRepository.map_utilisateur
